pkg/kubelet/images/image_manager_test.go

```diff
@@ -202,3 +206,172 @@ func TestApplyDefaultImageTag(t *testing.T) {
+			inspectErr:     nil,
+			pullerErr:      nil,
+			asyncExpected:  pullerExpects{[]string{"GetImageRef", "PullImage"}, nil},
+			expected: []pullerExpects{
+				{[]string{}, nil},
+			}},
+
+		{ // #1 image present, don't pull
+			containerImage: "present_image",
+			policy:         v1.PullIfNotPresent,
+			inspectErr:     nil,
+			pullerErr:      nil,
+			asyncExpected:  pullerExpects{[]string{"GetImageRef"}, nil},
+			expected: []pullerExpects{
+				{[]string{}, nil},
+				{[]string{"GetImageRef"}, nil},
+			}},
+		// #2 image present, pull it
+		{containerImage: "present_image",
+			policy:        v1.PullAlways,
+			inspectErr:    nil,
+			pullerErr:     nil,
+			asyncExpected: pullerExpects{[]string{"GetImageRef", "PullImage"}, nil},
+			expected: []pullerExpects{
+				{[]string{}, nil},
+				{[]string{"GetImageRef", "PullImage"}, nil},
+				{[]string{"GetImageRef", "PullImage"}, nil},
+			}},
+		// #3 missing image, error PullNever
+		{containerImage: "missing_image",
+			policy:        v1.PullNever,
+			inspectErr:    nil,
+			pullerErr:     nil,
+			asyncExpected: pullerExpects{[]string{"GetImageRef"}, ErrImageNeverPull},
+			expected: []pullerExpects{
+				{[]string{"GetImageRef"}, ErrImageNeverPull},
+				{[]string{"GetImageRef"}, ErrImageNeverPull},
+				{[]string{"GetImageRef"}, ErrImageNeverPull},
+			}},
+		// #4 missing image, unable to inspect
+		{containerImage: "missing_image",
+			policy:        v1.PullIfNotPresent,
+			inspectErr:    errors.New("unknown inspectError"),
+			pullerErr:     nil,
+			asyncExpected: pullerExpects{[]string{"GetImageRef"}, ErrImageInspect},
+			expected: []pullerExpects{
+				{[]string{"GetImageRef"}, ErrImageInspect},
+				{[]string{"GetImageRef"}, ErrImageInspect},
+				{[]string{"GetImageRef"}, ErrImageInspect},
+			}},
+		// #5 missing image, unable to fetch
+		{containerImage: "typo_image",
+			policy:        v1.PullIfNotPresent,
+			inspectErr:    nil,
+			pullerErr:     errors.New("404"),
+			asyncExpected: pullerExpects{[]string{"GetImageRef", "PullImage"}, nil},
+			expected: []pullerExpects{
+				{[]string{}, ErrImagePull},
+				{[]string{"GetImageRef", "PullImage"}, ErrImagePull},
+				{[]string{"GetImageRef"}, ErrImagePullBackOff},
+				{[]string{"GetImageRef", "PullImage"}, ErrImagePull},
+				{[]string{"GetImageRef"}, ErrImagePullBackOff},
+				{[]string{"GetImageRef"}, ErrImagePullBackOff},
+			}},
+	}
+}
+
+var _ imagePuller = &immediateTestPuller{}
+
+type immediateTestPuller struct {
+	imageService kubecontainer.ImageService
+}
+
+func (pip *immediateTestPuller) pullImage(spec kubecontainer.ImageSpec, pullSecrets []v1.Secret, pullChan chan<- pullResult, podSandboxConfig *runtimeapi.PodSandboxConfig) {
+	imageRef, err := pip.imageService.PullImage(spec, pullSecrets, podSandboxConfig)
+	pullChan <- pullResult{
+		imageRef: imageRef,
+		err:      err,
+	}
+	return
+}
+
+func setImagePuller(manager ImageManager, puller imagePuller) {
+	manager.(*imageManager).puller = puller
```

Comment:
Wasn't sure how best to handle this need to switch the `imagePuller`, I played with adding a `setPuller` onto the `ImageManager` interface but wasn't sure if that felt a bit unnecessary as I can do this cast. Would be good to get thoughts on other approaches on this one. Comment on line 317 goes into detail on why it's needed.